engine/src/main/java/org/hibernate/search/engine/backend/index/spi/IndexManagerState.java

```diff
@@ -0,0 +1,17 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.engine.backend.index.spi;
+
+/**
+ * A state from an existing {@link IndexManagerImplementor}
+ * that can be used to create a new instance of it.
+ */
+public interface IndexManagerState {
```

Comment:
Considering we'll potentially need to save state anywhere (backend, index manager, shard holder, shard, ...), and state will consist in potentially many types of values, I'd go with a more generic abstraction. Something like this:

```java
public class SavedState {

    private static final SavedState EMPTY = new SavedState.Builder().build();

    public static SavedState empty() {
        return EMPTY;
    }

    private final Map<Key<?>, Object> content;

    private SavedState(Builder builder) {
        this.content = builder.content;
    }

    public <T> Optional<T> get(Key<T> key, T defaultValue) {
        return Optional.ofNullable( (T) content.get( key ) );
    }

    public static <T> Key<T> key(String name) {
        Contracts.assertNotNullNorEmpty( name, "name" );
        return new Key<>( name );
    }

    public final class Key<T> {

        private final String name;

        private Key(String name) {
            this.name = name;
        }

        @Override
        public String toString() {
            return getClass().getSimpleName() + "[" + name + "]";
        }

        @Override
        public boolean equals(Object o) {
            if ( this == o ) {
                return true;
            }
            if ( o == null || getClass() != o.getClass() ) {
                return false;
            }
            Key<?> that = (Key<?>) o;
            return Objects.equals( name, that.name );
        }

        @Override
        public int hashCode() {
            return Objects.hash( name );
        }

    }

    public static Builder builder() {
        return new Builder();
    }

    public static final class Builder {
        private final Map<Key<?>, Object> content = new HashMap<>();

        private final Builder() {
        }

        public <T> Builder put(Key<T> key, T value) {
            content.put(key, value);
        }

        public SavedState build() {
            return new SavedState( this );
        }
    }

}
```

Then usage would be like this:

```java
public class LuceneIndexManagerImpl {
    private static SavedState.Key<SavedState> SHARD_HOLDER_KEY = SavedState.key("shard_holder");

    public SavedState saveForRestart() {
        return SavedState.builder()
                .put( SHARD_HOLDER_KEY, shardHolder.save() )
                // ...
                .build();
    }

    public void preStart(SavedState state) {
        shardHolder.preStart( state.get( SHARD_HOLDER_KEY ).orElse( SavedState.EMPTY ) );
    }
}
```

```java
public class Shard {
    private static SavedState.Key<DirectoryHolder> DIRECTORY_HOLDER_KEY = SavedState.key("directory_holder");

    public SavedState saveForRestart() {
        try {
            return SavedState.builder()
                    .put( DIRECTORY_HOLDER_KEY, directoryHolder )
                    .build();
        }
        finally {
            savedForRestart = true;
        }
    }

    public void preStart(SavedState state) {
        directoryHolder = state.get( DIRECTORY_HOLDER_KEY ).orElse( null );
        if ( directoryHolder == null ) {
           // ... create directory holder ...
        }
    }

    public void close() {
        if (!savedForRestart) {
            directoryHolder.close();
        }
    }
}
```

For maps or lists of values, you can probably make do with something like this:

```java
    private static SavedState.Key<Map<Integer, Shard>> SHARDS = SavedState.key("shards");
```